kmer/libutil/md5lib/md5.1.txt



MD5(1)                   USER COMMANDS                     MD5(1)



NAME
     md5 - calculate a message-digest fingerprint (checksum)  for
     a file

SYNOPSIS
     md5 [ -t | -x | -sstring | filename(s) ]

DESCRIPTION
     md5 takes as input a message of arbitrary  length  and  pro-
     duces  as output a 128-bit "fingerprint" or "message digest"
     of the input.  It is conjectured that it is  computationally
     infeasible  to  produce two messages having the same message
     digest, or to produce any message having a  given  prespeci-
     fied  target  message digest.  The MD5 algorithm is intended
     for digital signature applications, where a large file  must
     be  "compressed"  in  a secure manner before being encrypted
     with a private (secret) key under a public-key  cryptosystem
     such as _R_S_A.

OPTIONS
     The following four options may be used in  any  combination,
     except that filename(s) must be the last objects on the com-
     mand line.

     -sstring prints a checksum of the given "string".

     -t runs a built-in time trial.

     -x runs a built-in test script.

     filename(s) prints a checksum(s) for each of the files.

SEE ALSO
     sum(1)

     RFC 1321 describes in detail the MD2, MD4, and MD5  message-
     digest algorithms.

ACKNOWLEDGEMENTS
     This program is placed in the public domain for free general
     use by RSA Data Security.














Sun Release 4.1     Last change: Feb 14, 1994                   1





src/CHANGES
The intent of this file it to document the _effect_ of recent code
changes, not just the change itself, which should be in the cvs log.

Changes can be specified by a noting a single file/version that is in
the commit or a tag.  If possible, use "cvs log" to get the date/time
that cvs thinks the commit was made, as it is different from local
time.

The comment should be a description of what you thinnk should be
changing, or just what really does change.

For searching convenience, please use BUGFIX when a bug is fixed, and
BUGGY when something breaks.

----------------------------------------
20060823 TAG: VERSION-3_20

----------------------------------------
20070119 TAG: BEFORE_MSG_DEL_1

Lots and lots of consensus changes since 3.20.

----------------------------------------
20070202 TAG: BEFORE_MSG_DEL_2

BUGFIX: Surrogates with more than two instances are now placed
correctly.  This results in more fragments in surrogates being placed
in contigs.  Results up to resolveSurrogates should be identical.

----------------------------------------
20070211 TAG: PRE-3_30-BRANCH (and VERSION-3_30)

VERSION-3_30 has some backported bug fixes, but it was branched here.
PRE-3_30-BRANCH is identical to the last tag.

----------------------------------------
20070221 TAG: TMP-3_31-STABLE_21FEB07

BUGGY!  InputDataTypes_CGW.h and GraphCGW_T.c removed the DEBUG_DATA
conditional, which changed the initial labeling of "potentially bogus"
mates, and changed scaffolding.

----------------------------------------
20070223 TAG: TMP-3_32-STABLE_23FEB07

Still buggy, the same as 3.31.

----------------------------------------
20070225 TAG: TMP-3_33-STABLE_25FEB07

Still buggy, the same as 3.31.

----------------------------------------
20070301 TAG: TMP-3_34-STABLE_03MAR07

Still buggy, the same as 3.31.

----------------------------------------
20070309 TAG: TMP-3_35-STABLE_09MAR07

BUGFIX!  Fixes the bug in 3.31, this is now the same as 3.30.

A change to the Chi Suared test results in more unitig link messages.

----------------------------------------
20070312 TAG: TMP-3_36-STABLE_12MAR07 (where it should be)

1) Consensus makes sure that identical (ungapped) reads are aligned
the same.  This changes both post-unitigger and post-scaffolding
consensus sequences.

2) The chimera detector now doesn't use evidence from OBT-deleted
reads.  This can slightly change the trimming on a small number of
reads.

----------------------------------------
20070314 TAG: TMP-3_37-STABLE_14MAR07

1) Overlap error rates now have 2 digits of precision, up from 1.
This can remove some borderline overlaps, and will change unitigger.

2) BUGGY!  Version 1.8 of AS_OVS/AS_OVS_overlapStore.c is needed to
fix it (from 3/16).

----------------------------------------
20070316 18:56:09 1.135 gdenisov AS_CNS/MultiAlignment_CNS.c
20070316 23:52:40 1.26  gdenisov AS_CNS/MultiAlignment_CNS.h

Local phasing.  Unknown what changes.

----------------------------------------


src/AS_UTL/md5.C
#include "md5.H"

//  The RSA MD5 implementation.  Functions md5_* (at the end) are glue
//  to kmer libutil.

//  See RFC1321, "The MD5 Message-Digest Algorithm", R. Rivest.

//  Copyright (C) 1991-2, RSA Data Security, Inc. Created 1991. All
//  rights reserved.
//
//  License to copy and use this software is granted provided that it
//  is identified as the "RSA Data Security, Inc. MD5 Message-Digest
//  Algorithm" in all material mentioning or referencing this software
//  or this function.
//
//  License is also granted to make and use derivative works provided
//  that such works are identified as "derived from the RSA Data
//  Security, Inc. MD5 Message-Digest Algorithm" in all material
//  mentioning or referencing the derived work.
//
//  RSA Data Security, Inc. makes no representations concerning either
//  the merchantability of this software or the suitability of this
//  software for any particular purpose. It is provided "as is"
//  without express or implied warranty of any kind.
//
//  These notices must be retained in any copies of any part of this
//  documentation and/or software.


typedef struct {
  uint32  state[4];                                   // state (ABCD)
  uint32  count[2];        // number of bits, modulo 2^64 (lsb first)
  unsigned char buffer[64];                           // input buffer
} MD5_CTX;


static void MD5Init(MD5_CTX *);
static void MD5Update(MD5_CTX *, unsigned char const *, size_t);
static void MD5Final(unsigned char [16], MD5_CTX *);

static void MD5Transform(uint32 [4], unsigned char const [64]);
static void Encode(unsigned char *, uint32 *, unsigned int);
static void Decode(uint32 *, unsigned char const *, unsigned int);

//  Constants for MD5Transform routine.
#define S11 7
#define S12 12
#define S13 17
#define S14 22
#define S21 5
#define S22 9
#define S23 14
#define S24 20
#define S31 4
#define S32 11
#define S33 16
#define S34 23
#define S41 6
#define S42 10
#define S43 15
#define S44 21

static unsigned char PADDING[64] = {
  0x80, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
};

//  F, G, H and I are basic MD5 functions.
#define F(x, y, z) (((x) & (y)) | ((~x) & (z)))
#define G(x, y, z) (((x) & (z)) | ((y) & (~z)))
#define H(x, y, z) ((x) ^ (y) ^ (z))
#define I(x, y, z) ((y) ^ ((x) | (~z)))

//  ROTATE_LEFT rotates x left n bits.
#define ROTATE_LEFT(x, n) (((x) << (n)) | ((x) >> (32-(n))))

//  FF, GG, HH, and II transformations for rounds 1, 2, 3, and 4.
//  Rotation is separate from addition to prevent recomputation.
#define FF(a, b, c, d, x, s, ac) { \
 (a) += F ((b), (c), (d)) + (x) + (uint32)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }
#define GG(a, b, c, d, x, s, ac) { \
 (a) += G ((b), (c), (d)) + (x) + (uint32)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }
#define HH(a, b, c, d, x, s, ac) { \
 (a) += H ((b), (c), (d)) + (x) + (uint32)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }
#define II(a, b, c, d, x, s, ac) { \
 (a) += I ((b), (c), (d)) + (x) + (uint32)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }

//  MD5 initialization. Begins an MD5 operation, writing a new context.
//
void MD5Init (MD5_CTX *context) {
  context->count[0] = context->count[1] = 0;
  //  Load magic initialization constants.
  context->state[0] = 0x67452301;
  context->state[1] = 0xefcdab89;
  context->state[2] = 0x98badcfe;
  context->state[3] = 0x10325476;
}

//  MD5 block update operation. Continues an MD5 message-digest
//  operation, processing another message block, and updating the
//  context.
//
void MD5Update (MD5_CTX *context, unsigned char const *input, size_t inputLen) {
  unsigned int i, index, partLen;

  //  Compute number of bytes mod 64
  index = (unsigned int)((context->count[0] >> 3) & 0x3F);

  //  Update number of bits
  if ((context->count[0] += ((uint32)inputLen << 3))
      < ((uint32)inputLen << 3))
    context->count[1]++;
  context->count[1] += ((uint32)inputLen >> 29);

  partLen = 64 - index;

  //  Transform as many times as possible.
  if (inputLen >= partLen) {
    memcpy(&context->buffer[index], input, partLen);
    MD5Transform(context->state, context->buffer);

    for (i = partLen; i + 63 < inputLen; i += 64)
      MD5Transform(context->state, &input[i]);

    index = 0;
  }
  else
    i = 0;

  //  Buffer remaining input
  memcpy(&context->buffer[index], &input[i], inputLen-i);
}

//  MD5 finalization. Ends an MD5 message-digest operation, writing the
//  the message digest and zeroizing the context.
//
void MD5Final (unsigned char digest[16], MD5_CTX *context) {
  unsigned char bits[8];
  unsigned int index, padLen;

  //  Save number of bits
  Encode (bits, context->count, 8);

  //  Pad out to 56 mod 64.
  index = (unsigned int)((context->count[0] >> 3) & 0x3f);
  padLen = (index < 56) ? (56 - index) : (120 - index);
  MD5Update (context, PADDING, padLen);

  //  Append length (before padding)
  MD5Update (context, bits, 8);
  //  Store state in digest
  Encode (digest, context->state, 16);

  //  Zeroize sensitive information.
  memset(context, 0, sizeof(*context));
}

//  MD5 basic transformation. Transforms state based on block.
//
static void MD5Transform(uint32 state[4], unsigned char const block[64]) {
  uint32 a = state[0], b = state[1], c = state[2], d = state[3], x[16];

  Decode(x, block, 64);

  //  Round 1
  FF (a, b, c, d, x[ 0], S11, 0xd76aa478); //  1
  FF (d, a, b, c, x[ 1], S12, 0xe8c7b756); //  2
  FF (c, d, a, b, x[ 2], S13, 0x242070db); //  3
  FF (b, c, d, a, x[ 3], S14, 0xc1bdceee); //  4
  FF (a, b, c, d, x[ 4], S11, 0xf57c0faf); //  5
  FF (d, a, b, c, x[ 5], S12, 0x4787c62a); //  6
  FF (c, d, a, b, x[ 6], S13, 0xa8304613); //  7
  FF (b, c, d, a, x[ 7], S14, 0xfd469501); //  8
  FF (a, b, c, d, x[ 8], S11, 0x698098d8); //  9
  FF (d, a, b, c, x[ 9], S12, 0x8b44f7af); //  10
  FF (c, d, a, b, x[10], S13, 0xffff5bb1); //  11
  FF (b, c, d, a, x[11], S14, 0x895cd7be); //  12
  FF (a, b, c, d, x[12], S11, 0x6b901122); //  13
  FF (d, a, b, c, x[13], S12, 0xfd987193); //  14
  FF (c, d, a, b, x[14], S13, 0xa679438e); //  15
  FF (b, c, d, a, x[15], S14, 0x49b40821); //  16

  //  Round 2
  GG (a, b, c, d, x[ 1], S21, 0xf61e2562); //  17
  GG (d, a, b, c, x[ 6], S22, 0xc040b340); //  18
  GG (c, d, a, b, x[11], S23, 0x265e5a51); //  19
  GG (b, c, d, a, x[ 0], S24, 0xe9b6c7aa); //  20
  GG (a, b, c, d, x[ 5], S21, 0xd62f105d); //  21
  GG (d, a, b, c, x[10], S22,  0x2441453); //  22
  GG (c, d, a, b, x[15], S23, 0xd8a1e681); //  23
  GG (b, c, d, a, x[ 4], S24, 0xe7d3fbc8); //  24
  GG (a, b, c, d, x[ 9], S21, 0x21e1cde6); //  25
  GG (d, a, b, c, x[14], S22, 0xc33707d6); //  26
  GG (c, d, a, b, x[ 3], S23, 0xf4d50d87); //  27
  GG (b, c, d, a, x[ 8], S24, 0x455a14ed); //  28
  GG (a, b, c, d, x[13], S21, 0xa9e3e905); //  29
  GG (d, a, b, c, x[ 2], S22, 0xfcefa3f8); //  30
  GG (c, d, a, b, x[ 7], S23, 0x676f02d9); //  31
  GG (b, c, d, a, x[12], S24, 0x8d2a4c8a); //  32

  //  Round 3
  HH (a, b, c, d, x[ 5], S31, 0xfffa3942); //  33
  HH (d, a, b, c, x[ 8], S32, 0x8771f681); //  34
  HH (c, d, a, b, x[11], S33, 0x6d9d6122); //  35
  HH (b, c, d, a, x[14], S34, 0xfde5380c); //  36
  HH (a, b, c, d, x[ 1], S31, 0xa4beea44); //  37
  HH (d, a, b, c, x[ 4], S32, 0x4bdecfa9); //  38
  HH (c, d, a, b, x[ 7], S33, 0xf6bb4b60); //  39
  HH (b, c, d, a, x[10], S34, 0xbebfbc70); //  40
  HH (a, b, c, d, x[13], S31, 0x289b7ec6); //  41
  HH (d, a, b, c, x[ 0], S32, 0xeaa127fa); //  42
  HH (c, d, a, b, x[ 3], S33, 0xd4ef3085); //  43
  HH (b, c, d, a, x[ 6], S34,  0x4881d05); //  44
  HH (a, b, c, d, x[ 9], S31, 0xd9d4d039); //  45
  HH (d, a, b, c, x[12], S32, 0xe6db99e5); //  46
  HH (c, d, a, b, x[15], S33, 0x1fa27cf8); //  47
  HH (b, c, d, a, x[ 2], S34, 0xc4ac5665); //  48

  //  Round 4
  II (a, b, c, d, x[ 0], S41, 0xf4292244); //  49
  II (d, a, b, c, x[ 7], S42, 0x432aff97); //  50
  II (c, d, a, b, x[14], S43, 0xab9423a7); //  51
  II (b, c, d, a, x[ 5], S44, 0xfc93a039); //  52
  II (a, b, c, d, x[12], S41, 0x655b59c3); //  53
  II (d, a, b, c, x[ 3], S42, 0x8f0ccc92); //  54
  II (c, d, a, b, x[10], S43, 0xffeff47d); //  55
  II (b, c, d, a, x[ 1], S44, 0x85845dd1); //  56
  II (a, b, c, d, x[ 8], S41, 0x6fa87e4f); //  57
  II (d, a, b, c, x[15], S42, 0xfe2ce6e0); //  58
  II (c, d, a, b, x[ 6], S43, 0xa3014314); //  59
  II (b, c, d, a, x[13], S44, 0x4e0811a1); //  60
  II (a, b, c, d, x[ 4], S41, 0xf7537e82); //  61
  II (d, a, b, c, x[11], S42, 0xbd3af235); //  62
  II (c, d, a, b, x[ 2], S43, 0x2ad7d2bb); //  63
  II (b, c, d, a, x[ 9], S44, 0xeb86d391); //  64

  state[0] += a;
  state[1] += b;
  state[2] += c;
  state[3] += d;

  //  Zeroize sensitive information.
  memset (x, 0, sizeof(x));
}

//  Encodes input (uint32) into output (unsigned char). Assumes len is
//  a multiple of 4.
//
static void Encode (unsigned char *output, uint32 *input, unsigned int len) {
  unsigned int i, j;

  for (i = 0, j = 0; j < len; i++, j += 4) {
    output[j] = (unsigned char)(input[i] & 0xff);
    output[j+1] = (unsigned char)((input[i] >> 8) & 0xff);
    output[j+2] = (unsigned char)((input[i] >> 16) & 0xff);
    output[j+3] = (unsigned char)((input[i] >> 24) & 0xff);
  }
}

//  Decodes input (unsigned char) into output (uint32). Assumes len is
//  a multiple of 4.
//
static void Decode (uint32 *output, unsigned char const *input, unsigned int len) {
  unsigned int i, j;

  for (i = 0, j = 0; j < len; i++, j += 4)
    output[i] = ((uint32)input[j]) | (((uint32)input[j+1]) << 8) |
      (((uint32)input[j+2]) << 16) | (((uint32)input[j+3]) << 24);
}


////////////////////////////////////////////////////////////////////////////////
//
//  kmer glue functions
//
////////////////////////////////////////////////////////////////////////////////


int
md5_compare(void const *a, void const *b) {
  md5_s const *A = (md5_s const *)a;
  md5_s const *B = (md5_s const *)b;

  if (A->a < B->a) return(-1);
  if (A->a > B->a) return(1);
  if (A->b < B->b) return(-1);
  if (A->b > B->b) return(1);
  return(0);
}

static const char *md5_letters = "0123456789abcdef";

char*
md5_toascii(md5_s *m, char *s) {
  int i;
  for (i=0; i<16; i++) {
    s[15-i   ] = md5_letters[(m->a >> 4*i) & 0x0f];
    s[15-i+16] = md5_letters[(m->b >> 4*i) & 0x0f];
  }
  s[32] = 0;

  return(s);
}

md5_s*
md5_string(md5_s *m, char *s, uint32 l) {
  MD5_CTX         ctx;
  unsigned char   dig[16];
  int             i = 0;

  if (m == NULL) {
    errno = 0;
    m = new md5_s;
    if (errno) {
      fprintf(stderr, "md5_string()-- Can't allocate a md5_s.\n%s\n", strerror(errno));
      exit(1);
    }
  }

  MD5Init(&ctx);
  MD5Update(&ctx, (unsigned char*)s, l);
  MD5Final(dig, &ctx);

  m->a = dig[0];
  while (i<8) {
    m->a <<= 8;
    m->a |= dig[i++];
  }

  m->b  = dig[i++];
  while (i<16) {
    m->b <<= 8;
    m->b |= dig[i++];
  }

  return(m);
}

static
md5_increment_s*
md5_increment_initialize(void) {
  md5_increment_s *m;

  errno = 0;
  m = new md5_increment_s;
  if (errno) {
    fprintf(stderr, "md5_increment_*()-- Can't allocate a md5_increment_s.\n%s\n", strerror(errno));
    exit(1);
  }

  m->context = new MD5_CTX;
  if (errno) {
    fprintf(stderr, "md5_increment_*()-- Can't allocate a md5 context.\n%s\n", strerror(errno));
    exit(1);
  }
  MD5Init((MD5_CTX *)m->context);

  m->bufferPos = 0;

  return(m);
}

md5_increment_s*
md5_increment_char(md5_increment_s *m, char s) {

  if (m == NULL)
    m = md5_increment_initialize();

  m->buffer[m->bufferPos++] = s;

  if (m->bufferPos == MD5_BUFFER_SIZE) {
    MD5Update((MD5_CTX *)m->context, m->buffer, m->bufferPos);
    m->bufferPos = 0;
  }

  return(m);
}

md5_increment_s*
md5_increment_block(md5_increment_s *m, char *s, uint32 l) {

  if (m == NULL)
    m = md5_increment_initialize();

  MD5Update((MD5_CTX *)m->context, (unsigned char*)s, l);

  return(m);
}

void
md5_increment_finalize(md5_increment_s *m) {
  MD5_CTX        *ctx = (MD5_CTX *)m->context;
  unsigned char   dig[16];
  int             i = 0;

  if (m->bufferPos > 0) {
    MD5Update((MD5_CTX *)m->context, m->buffer, m->bufferPos);
    m->bufferPos = 0;
  }

  MD5Final(dig, ctx);

  m->a = dig[0];
  while (i<8) {
    m->a <<= 8;
    m->a |= dig[i++];
  }

  m->b  = dig[i++];
  while (i<16) {
    m->b <<= 8;
    m->b |= dig[i++];
  }

  m->context = 0L;

  delete ctx;
}

void
md5_increment_destroy(md5_increment_s *m) {
  delete m;
}
